samza-core/src/main/java/org/apache/samza/zk/ZkUtils.java

```diff
@@ -438,11 +442,48 @@ public void publishJobModelVersion(String oldVersion, String newVersion) {
         "(actual data version after update = " + stat.getVersion() + ")");
   }
 
+  // validate Zk protocol currently used is the same as in this participant
+  private void validateZkVersion() {
+    String rootPath = keyBuilder.getRootPath();
+    if (!zkClient.exists(rootPath)) {
```

Comment:
I would recommend moving Line 447-458 to a separate function called validateRootPath since this is not for version validation. I think it is better to have single responsibility for each function.